### ISSUE TITLE
Index out of range exception in FilteredReadOnlyObservableCollection

### ISSUE THREAD

**Issue by nanoctr** [runceel/ReactiveProperty#379]
First of all, I'd like to thank you for building this library. We use it in our company for a WPF application and it's been really helpful :)

Unfortunately, we discovered a small bug in the implementation of `FilteredReadOnlyObservableCollection`. Under certain circumstances, it throws index out of range exceptions when adding elements to the underlying source collection.

## Exception message
```
System.ArgumentOutOfRangeException: Index was out of range. Must be non-negative and less than the size of the collection. (Parameter 'index')
    at System.Collections.Generic.List`1.get_Item(Int32 index)
   at Reactive.Bindings.Helpers.FilteredReadOnlyObservableCollection`3.SourceElementChanged(TElement x) in C:\Dev\ReactiveProperty\Source\ReactiveProperty.NETStandard\Helpers\FilteredReadOnlyObservableCollection.cs:line 110
(...)
```

## Steps to reproduce
The following code exhibits the faulty behavior:
```
public class OutOfRangeErrorTest
{
    // Doesn't actually matter, this just exists so we can use `ObserveProperty()` further down
    class Foo : INotifyPropertyChanged
    {
        public int Bar => 0;
        
        public event PropertyChangedEventHandler PropertyChanged;
    }

    public void ThrowsError()
    {
        var observableCollection = new ObservableCollection<Foo>();

        var filteredCollection = observableCollection.ToFilteredReadOnlyObservableCollection(
            filter: foo => foo.Bar > 10,
            elementStatusChangedFactory: foo => foo.ObserveProperty(x => x.Bar));

        var element = new Foo();
        observableCollection.Add(element); // Exception thrown here
    }
}
```

I took a look at the code in https://github.com/runceel/ReactiveProperty/blob/main/Source/ReactiveProperty.NETStandard/Helpers/FilteredReadOnlyObservableCollection.cs and ran it through debugging locally.
It seems that `SourceElementChanged` runs _before_ `Source_CollectionChanged` in this case. So, when adding a new element, `SourceElementChanged` is executed first. `IndexList` doesn't yet contain the new element in line 110... and there you go, that's our culprit :)

I guess our use of `ObserveProperty()` is problematic for some reason.

## Fix

A simple fix is to register the event handler for `source.CollectionChanged` _before_ setting up the `SourceElementChanged` subscription.
So, the code from line 89 onwards would be:
```
// collection changed(support single changed only)
source.CollectionChanged += Source_CollectionChanged;

// propertychanged
CollectionUtilities.ObserveElementCore<TCollection, TElement, TElement>(source, (x, observer) =>
    (...)
```
With this change, CollectionChanged is always executed before the elementChanged observable.

But I'm not sure if this is really the right fix. I don't know enough about your codebase to be sure, so I didn't want to open a PR straight away :)

If there's anything else I can do to help, please let me know!

**Comment by runceel** [comment]
Hi @nanoctr 

Thank you for reporting this issue.
I have released v8.1.1 that is fixed this issue.

Your approach is correct way I thought. so I did that. :)

**Comment by runceel** [comment]
The release page: https://github.com/runceel/ReactiveProperty/releases/tag/8.1.1
NuGet: https://www.nuget.org/packages/ReactiveProperty/8.1.1
